core/modules/url_filter.cc

```diff
@@ -0,0 +1,279 @@
+
+  return pkt;
+}
+
+// Generate a TCP RST packet
+inline static bess::Packet *GenerateResetPacket(
+    const EthHeader::Address &src_eth, const EthHeader::Address &dst_eth,
+    uint32_t src_ip, uint32_t dst_ip, uint16_t src_port, uint16_t dst_port,
+    uint32_t seq, uint32_t ack) {
+  bess::Packet *pkt = bess::Packet::Alloc();
+  char *ptr = static_cast<char *>(pkt->buffer()) + SNBUF_HEADROOM;
+  pkt->set_data_off(SNBUF_HEADROOM);
+  pkt->set_total_len(sizeof(rst_template));
+  pkt->set_data_len(sizeof(rst_template));
+
+  rte_memcpy(ptr, &rst_template, sizeof(rst_template));
+
+  EthHeader *eth = reinterpret_cast<EthHeader *>(ptr);
+  Ipv4Header *ip = reinterpret_cast<Ipv4Header *>(eth + 1);
+  // We know there is no IP option
+  TcpHeader *tcp = reinterpret_cast<TcpHeader *>(ip + 1);
+
+  eth->dst_addr = dst_eth;
+  eth->src_addr = src_eth;
+  ip->src = src_ip;
+  ip->dst = dst_ip;
+  tcp->src_port = src_port;
+  tcp->dst_port = dst_port;
+  tcp->seq_num = seq;
+  tcp->ack_num = ack;
+  tcp->checksum =
+      rte_ipv4_udptcp_cksum(reinterpret_cast<const ipv4_hdr *>(ip), tcp);
+  ip->checksum = rte_ipv4_cksum(reinterpret_cast<const ipv4_hdr *>(ip));
+
+  return pkt;
+}
+
+pb_error_t UrlFilter::Init(const bess::pb::UrlFilterArg &arg) {
+  for (const auto &url : arg.blacklist()) {
+    if (blacklist_.find(url.host()) == blacklist_.end()) {
+      blacklist_.emplace(url.host(), Trie());
+    }
+    Trie &trie = blacklist_.at(url.host());
+    trie.Insert(url.path());
+  }
+  return pb_errno(0);
+}
+
+pb_cmd_response_t UrlFilter::CommandAdd(const bess::pb::UrlFilterArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, Init(arg));
+  return response;
+}
+
+pb_cmd_response_t UrlFilter::CommandClear(const bess::pb::EmptyArg &) {
+  blacklist_.clear();
+  return pb_cmd_response_t();
+}
+
+void UrlFilter::ProcessBatch(bess::PacketBatch *batch) {
+  gate_idx_t igate = get_igate();
+
+  // Pass reverse traffic
+  if (igate == 1) {
+    RunChooseModule(1, batch);
+    return;
+  }
+
+  // Otherwise
+  bess::PacketBatch out_batches[2];
+  out_batches[0].clear();
+  out_batches[1].clear();
+  int cnt = batch->cnt();
+
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    struct EthHeader *eth = pkt->head_data<struct EthHeader *>();
+    struct Ipv4Header *ip = reinterpret_cast<struct Ipv4Header *>(eth + 1);
+
+    if (ip->protocol != 0x06) {
+      continue;
+    }
+
+    int ip_bytes = (ip->header_length & 0xf) << 2;
+    struct TcpHeader *tcp = reinterpret_cast<struct TcpHeader *>(
+        reinterpret_cast<uint8_t *>(ip) + ip_bytes);
+
+    Flow flow;
+    flow.src_ip = ip->src;
+    flow.dst_ip = ip->dst;
+    flow.src_port = tcp->src_port;
+    flow.src_port = tcp->dst_port;
+
+    // Check if the flow is already blocked
+    std::unordered_map<Flow, uint64_t, FlowHash>::iterator it;
+    if ((it = blocked_flows_.find(flow)) != blocked_flows_.end()) {
+      uint64_t now = tsc_to_ns(rdtsc());
```

Comment:
`rdtsc()` is an expensive operation. You can use `ctx.current_ns()`, which is refreshed for every scheduling round, in place of `tsc_to_ns(rdtsc())`